taichi/backends/vulkan/runtime.cpp

```diff
@@ -440,57 +447,88 @@ void VkRuntime::launch_kernel(KernelHandle handle, RuntimeContext *host_ctx) {
          /*export_sharing=*/false, AllocUsage::Storage});
   }
 
+  // Create context blitter
   auto ctx_blitter = HostDeviceContextBlitter::maybe_make(
       &ti_kernel->ti_kernel_attribs().ctx_attribs, host_ctx, device_,
       host_result_buffer_, ctx_buffer.get(), ctx_buffer_host.get());
 
   std::unordered_map<int, DeviceAllocation> ext_arrays;
+  std::unordered_map<int, size_t> ext_array_size;
 
+  // Prepare context buffers & arrays
+  const auto &args = ti_kernel->ti_kernel_attribs().ctx_attribs.args();
   if (ctx_blitter) {
     TI_ASSERT(ti_kernel->get_ctx_buffer_size());
 
-    const auto &args = ti_kernel->ti_kernel_attribs().ctx_attribs.args();
     int i = 0;
     for (auto &arg : args) {
       if (arg.is_array) {
-        if (arg.stride) {
+        if (host_ctx->is_device_allocation[i]) {
+          // NDArray
+          if (host_ctx->args[i]) {
+            ext_arrays[i] = *(DeviceAllocation *)(host_ctx->args[i]);
+          } else {
+            ext_arrays[i] = kDeviceNullAllocation;
+          }
+        } else {
+          // Compute ext arr sizes
+          size_t size = arg.stride;
+
+          for (int ax = 0; ax < 8; ax++) {
+            // FIXME: how and when do we determine the size of ext arrs?
```

Comment:
FYI opengl got it from `kernel->args[i].size` but IMHO we should remove that in `Args` class. https://github.com/taichi-dev/taichi/blob/master/taichi/program/callable.h#L22 
But the algorithm here also need to handle the a torch tensor with shape `[1, 0, 1]`  (total size is still 0).

Reply:
Can we unify the shape calculation algo here?